tee-worker/core/parentchain/indirect-calls-executor/src/indirect_calls_executor.rs

```diff
@@ -247,179 +246,43 @@ impl<ShieldingKeyRepository, StfEnclaveSigner, TopPoolAuthor, NodeMetadataProvid
-					)?;
-					let validation_data = ValidationData::decode(
-						&mut shielding_key.decrypt(&encrypted_validation_data).unwrap().as_slice(),
-					)?;
-
-					if let Some((multiaddress_account, _, _)) = xt.signature {
-						let account = AccountIdLookup::lookup(multiaddress_account)?;
-						let enclave_account_id = self.stf_enclave_signer.get_enclave_account()?;
-						let trusted_call = TrustedCall::verify_identity_preflight(
-							enclave_account_id,
-							account,
-							identity,
-							validation_data,
-							block_number
-								.try_into()
-								.map_err(|_| crate::error::Error::ConvertParentchainBlockNumber)?,
-						);
-						let signed_trusted_call =
-							self.stf_enclave_signer.sign_call_with_self(&trusted_call, &shard)?;
-						let trusted_operation =
-							TrustedOperation::indirect_call(signed_trusted_call);
-
-						let encrypted_trusted_call =
-							shielding_key.encrypt(&trusted_operation.encode())?;
-						self.submit_trusted_call(shard, encrypted_trusted_call);
-					}
-				}
-			}
-
+			let verify_identity = VerifyIdentity { block_number: parentchain_block_number };
 			// Found RequestVC extrinsic
-			if let Ok(xt) = ParentchainUncheckedExtrinsic::<RequestVCFn>::decode(
-				&mut encoded_xt_opaque.as_slice(),
-			) {
-				if self.is_request_vc_funciton(&xt.function.0) {
-					let (_, shard, assertion) = xt.function;
-					let shielding_key = self.shielding_key_repo.retrieve_key()?;
-					debug!("Requested VC Assertion {:?}", assertion);
-
-					if let Some((multiaddress_account, _, _)) = xt.signature {
-						let account = AccountIdLookup::lookup(multiaddress_account)?;
-						let enclave_account_id = self.stf_enclave_signer.get_enclave_account()?;
-
-						let trusted_call = TrustedCall::build_assertion(
-							enclave_account_id,
-							account,
-							assertion,
-							shard,
-						);
-						let signed_trusted_call =
-							self.stf_enclave_signer.sign_call_with_self(&trusted_call, &shard)?;
-						let trusted_operation =
-							TrustedOperation::indirect_call(signed_trusted_call);
-
-						let encrypted_trusted_call =
-							shielding_key.encrypt(&trusted_operation.encode())?;
-						self.submit_trusted_call(shard, encrypted_trusted_call);
-					}
+			let request_vc = RequestVC {};
+
+			let executors: Vec<
+				&dyn DecorateExecutor<
+					ShieldingKeyRepository,
+					StfEnclaveSigner,
+					TopPoolAuthor,
+					NodeMetadataProvider,
+				>,
+			> = vec![
+				&set_user_shielding_key,
+				&create_identity,
+				&remove_identity,
+				&verify_identity,
+				&request_vc,
+			];
+			for executor in executors {
+				match executor.decode_and_execute(self, &mut encoded_xt_opaque.as_slice()) {
+					Ok(ExecutionStatus::Success) | Ok(ExecutionStatus::Skip) => break,
+					Ok(ExecutionStatus::NextExecutor) => continue,
+					Err(e) => {
+						log::error!("fail to execute indirect_call. due to {:?} ", e);
+						// continue or return error???
+						continue
```

Comment:
return error, I think. because we decode first, then execute, the error is the execute error. if the decode fails, it will continue.

Reply:
After my thinking, I also agree to return error. Because the logic of the original code needs to be kept the same, when encountering an error, exit the function

Reply:
So strange, I found different logic about error handling in this function. See:
https://github.com/litentry/litentry-parachain/blob/2371278fe9e014c4111865480b5f86d22511896a/tee-worker/core/parentchain/indirect-calls-executor/src/indirect_calls_executor.rs#L224-L233

https://github.com/litentry/litentry-parachain/blob/2371278fe9e014c4111865480b5f86d22511896a/tee-worker/core/parentchain/indirect-calls-executor/src/indirect_calls_executor.rs#L255-L261

Maybe we should take the same handling as the first one.

Reply:
Do you mean it should have error'ed out or not continued with the next `if`? I actually think our handling is correct while integritee's is wrong.

Btw I saw they are pushing successfully processed calls to `executed_shielding_calls`, maybe we should do the same? Although these are just used to compute a merkle hash

Reply:
I think here it shouldn't return error. The code Zhiming cited is logic from us, which is not appropriate according to INTEGRITEE's intention.

Reply:
Sorry maybe I didn't make myself clear, I actually meant `break` here. We should stop processing of this extrinsic (= don't try the next executor) and continue with the next extrinsic. When all extrinsics are processed, we call `create_processed_parentchain_block_call` to report to the parachain.

And yes, we shouldn't error out using the `?`: `create_processed_parentchain_block_call` needs to be called in any case.